docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,137 @@
+incentive aligned mechanism. In addition, the lack of prioritization also leads
+to cascading effects in terms of DoS and various attack vectors on networks,
+e.g. [cosmos/cosmos-sdk#8224](https://github.com/cosmos/cosmos-sdk/discussions/8224).
+
+Thus, Tendermint Core needs the ability for an application and its users to
+prioritize transactions in a flexible and performant manner.
+
+## Alternative Approaches
+
+When considering which approach to take for a priority-based flexible and
+performant mempool, there are two core candidates. The first candidate in less
+invasive in the required  set of protocol and implementation changes, which
+simply extends the existing `CheckTx` ABCI method. The second candidate essentially
+involves the introduction of new ABCI method(s) and would require a higher degree
+of complexity in protocol and implementation changes, some of which may either
+overlap or conflict with the upcoming introduction of [ABCI++](https://github.com/tendermint/spec/blob/master/rfc/004-abci%2B%2B.md).
+
+For more information on the various approaches and proposals, please see the
+[mempool discussion](https://github.com/tendermint/tendermint/discussions/6295).
+
+## Decision
+
+To incorporate a priority-based flexible and performant mempool in Tendermint Core,
+we will introduce new fields, `priority`, `sender`, and `nonce` , into the
+`ResponseCheckTx` type and augment the existing mempool data structure to
+facilitate prioritization of uncommitted transactions in addition to extended
+functionality such as replace-by-priority and allowing multiple transactions to
+exist from the same sender with varying priorities.
+
+## Detailed Design
+
+### CheckTx
+
+We introduce the following new fields into the `ResponseCheckTx` type:
+
+```diff
+message ResponseCheckTx {
+  uint32         code       = 1;
+  bytes          data       = 2;
+  string         log        = 3;  // nondeterministic
+  string         info       = 4;  // nondeterministic
+  int64          gas_wanted = 5 [json_name = "gas_wanted"];
+  int64          gas_used   = 6 [json_name = "gas_used"];
+  repeated Event events     = 7 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string         codespace  = 8;
++ int64          priority   = 9;
++ string         sender     = 10;
++ int64          nonce      = 11;
+}
+```
+
+It is entirely up the application in determining how these fields are populated
+and with what values, e.g. the `sender` could be the signer  and fee payer 
+the transaction, the `priority` could be the cumulative sum of the fee(s), and
+the `nonce` could be the signer's sequence number/nonce.
+
+Only `sender` is required, while `priority` and `nonce` can be omitted which
+would result in using the default value of zero.
+
+### Mempool
+
+The existing concurrent-safe linked-list will be removed entirely in favor of a
+thread-safe map of `<sender:[]Tx>`, i.e a mapping from `sender` to a list of `Tx`
```

Comment:
how are senders ordered relative to eachother? it seems like we'd want to be able to have some kind of total ordering/heap, but there are lots 

- order senders relative to each other based on the number of transactions, and the total priority values of the transactions pending from that sender
- order transactions from each sender based on nonce/sequence and then priority

Reply:
> how are senders ordered relative to each other? 

There is no ordering between senders. We don't need it.

> it seems like we'd want to be able to have some kind of total ordering/heap

Why? The main ordering we care about is priority which already has an index.

> order senders relative to each other based on the number of transactions, and the total priority values of the transactions pending from that sender

These are not requirements of the mempool. Proposers do not care about total number of txs from a sender or anything else really. They simply care about extracting as much value as possible.

> order transactions from each sender based on nonce/sequence and then priority

Transactions are already ordered by nonce. There is a priority index built on top of this. This is very similar to [Diem's mempool design](https://github.com/diem/diem/tree/main/mempool#implementation-details).

Reply:
are sequence numbers globally unique or only unique per-sender? 

if the memory pool is a map of senders, rather than a totally ordered structure what method do we use to select from which sender the next transaction comes from?

Reply:
They are unique to the sender. From reading the design is to use two data structure, one to handle sender and sequence numbers and one to handle priority. This map wouldnt worry about the latter.

Reply:
> what method do we use to select from which sender the next transaction comes from?

The priority heap. Select the next highest priority tx from the heap, then check that its at the top of the list for the sender

Reply:
Cool cool, this makes sense to me now, though it might help other hapless readers in the future to put the "priority heap" part first and then put the sender-indexed map second in the document just for improved readability.

Reply:
Resolving as we're moving towards a single tx per signer 👍